Converting Result to Option: C error. This should never happen.

I wrote this tiny test program because I was confused about `Result` vs. `Option`.

Obviously the compiler feels the same :smiley: 


<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, "Fedora release 36 (Thirty Six)"
Processor: 4 cpus, 64bit, little endian, Intel(R) Xeon(R) CPU E3-1245 v3 @ 3.40GHz
CC version: cc (GCC) 12.1.1 20220507 (Red Hat 12.1.1-1)

getwd: /home/mfx/pp/lang-v/test-option-vs-result
vmodules: /home/mfx/.vmodules
vroot: /mnt/local/home-extra-storage-fast/home/mfx/code/github/markus-oberhumer-forks/vlang--v.git
vexe: /mnt/local/home-extra-storage-fast/home/mfx/code/github/markus-oberhumer-forks/vlang--v.git/vlang-mfx
vexe mtime: 2022-08-10 13:13:36
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.0 b0c32e0.1c6366e

Git version: git version 2.37.1
Git vroot status: weekly.2022.32-10-g1c6366eb
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 827f7452

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg bug.v`
```v
// Result
fn foo(a int) !int {
	return a + 1
}

// Option
fn bar(a int) ?int {
	return foo(a)
}

x := bar(1)?
println('$x')

```

**What did you expect to see?**

No internal compiler bug.


**What did you see instead?**
```
==================
/tmp/v_1000/bug.13898382030531097917.tmp.c:12167: error: cannot convert 'struct _result_int' to 'int'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/return.v
+++ b/vlib/v/checker/return.v
@@ -104,6 +104,12 @@ pub fn (mut c Checker) return_stmt(mut node ast.Return) {
 			c.error('cannot use `${c.table.type_to_str(got_typ)}` as type `${c.table.type_to_str(exp_type)}` in return argument',
 				pos)
 		}
+		if got_typ.has_flag(.result) && (!exp_type.has_flag(.result)
+			|| c.table.type_to_str(got_typ) != c.table.type_to_str(exp_type)) {
+			pos := node.exprs[expr_idxs[i]].pos()
+			c.error('cannot use `${c.table.type_to_str(got_typ)}` as type `${c.table.type_to_str(exp_type)}` in return argument',
+				pos)
+		}
 		if node.exprs[expr_idxs[i]] !is ast.ComptimeCall && !c.check_types(got_typ, exp_type) {
 			got_typ_sym := c.table.sym(got_typ)
 			mut exp_typ_sym := c.table.sym(exp_type)

--- a/vlib/v/checker/tests/return_result_type_mismatch.out
+++ b/vlib/v/checker/tests/return_result_type_mismatch.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/return_result_type_mismatch.vv:8:9: error: cannot use `!int` as type `?int` in return argument
+    6 | // Option
+    7 | fn bar(a int) ?int {
+    8 |     return foo(a)
+      |            ~~~~~~
+    9 | }
+   10 |

--- a/vlib/v/checker/tests/return_result_type_mismatch.vv
+++ b/vlib/v/checker/tests/return_result_type_mismatch.vv
@@ -1,0 +1,14 @@
+// Result
+fn foo(a int) !int {
+	return a + 1
+}
+
+// Option
+fn bar(a int) ?int {
+	return foo(a)
+}
+
+fn main() {
+	x := bar(1)?
+	println('$x')
+}

